src/Mod/TechDraw/Gui/QGIViewBalloon.cpp

```diff
@@ -489,33 +487,50 @@ void QGIViewBalloon::balloonLabelDragged(bool ctrl)
         }
     }
 
+    DrawView* balloonParent = getSourceView();
+    if (balloonParent)
+        // redraw the balloon at the new position
+        // note that we don't store the new position to the X/Y properties
+        // since the dragging is not yet finished
+        drawBalloon(true);
+
+    // store if origin is also moving to be able to later calc new origin and update feature
+    if (ctrl)
+        m_originDragged = true;
+}
+
+void QGIViewBalloon::balloonLabelDragFinished()
+{
+    // stores the final drag position for undo
+
+    auto dvb(dynamic_cast<TechDraw::DrawViewBalloon*>(getViewObject()));
+    if (!dvb)
+        return;
+
     double scale = 1.0;
     DrawView* balloonParent = getSourceView();
-    if (balloonParent != nullptr) {
+    if (balloonParent)
         scale = balloonParent->getScale();
-    }
```

Comment:
this can be rewritten as:
```c++
    if (const auto balloonParent = getSourceView())
        scale = balloonParent->getScale();
```

Reply:
This is a bugfix PR to be committed for 0.19 to fix an essential Techdraw feature - being able to undo changes.

I only make the code beautification because chennes recommended it. There are some more possible code improvements but not in this bugfix. Please make a new PR for further improvements - for 0.20.